File: C version/link/reverseList.c
#include<stdio.h>
#include<stdlib.h>
struct ListNode {
    int val;
    struct ListNode *next;
};

struct ListNode* reverseList(struct ListNode* head){
    struct ListNode* prev = NULL;
    struct ListNode* next = head;
    while(head!=NULL)
    {
        head = head->next;
        next->next = prev;
        prev = next;
        next = head;
    }
    printf("%d\n",prev->val);
    return prev;
}
void main(){
    int a[5] = {1,2,3,4,5};
    struct ListNode* head = NULL;
    struct ListNode* cur = NULL;
    int i = 0;
    for(;i<5;i++){
        if(i==0){
            struct ListNode* node = (struct ListNode*)malloc(sizeof(struct ListNode));
            node->val = a[i];
            node->next = (struct ListNode*)malloc(sizeof(struct ListNode));
            head = node;
            printf("%d\n",node->val);
            cur = node->next;
        }else{
            cur->val = a[i];
            cur->next = i==4?NULL:(struct ListNode*)malloc(sizeof(struct ListNode));
            printf("%d\n",cur->val);
            cur = cur->next;
        }
    }
    reverseList(head);
    // printf("%d\n",head->val);
    // printf("%d\n",head->next->val);
    // printf("%d\n",head->next->next->val);
    // printf("%d\n",head->next->next->next->val);
    // printf("%d\n",head->next->next->next->next->val);
    // printf("%d\n",head->next->next->next->next->next);
    getchar();
}

File: LinkedList/reverseList.js
function ListNode(val, next) {
    this.val = (val===undefined ? 0 : val)
    this.next = (next===undefined ? null : next)
}
function GenerateLink(list){
  let head = Object.create(null);
  let node = Object.create(null);
  for(let i=0;i<list.length;i++){
      if(i===0){
        head = new ListNode(list[i],null);
        node = head;
      }else{
        node.next = new ListNode(list[i],null);
        node = node.next;
      }
  }
  return head;
}
/**
* @param {ListNode} head
* @return {ListNode}
*/
var reverseList = function(head) {
    let tail = null;
    let shead = null;
    while(head){
        shead = new ListNode(head.val,null);
        if(tail){
            shead.next = tail;
        }
        tail = shead;
        head = head.next;
    }
    console.log(shead);
    return shead;
};
var reverseList = function(head) {//迭代法
    let pre = null;
    while(head!==null){
        const next = head.next;
        head.next = pre;
        pre = head;
        head = next;
    }
    return pre;
};
var reverseList = function(head) {//递归法
    function reverse(pre,cur){
        if(cur===null){
            return pre;
        }
        const next = cur.next;
        cur.next = pre;
        return reverse(cur,next);
    }
    return reverse(null,head);
}
console.log(reverseList(GenerateLink([1,2,3,4,5])))